### PR TITLE
763 msa precip distribution

### PR DESCRIPTION
**Pull request by msahn**
This is to implement precip distribution metrics into the PMP.

### CODE REVIEW

**Comment by msahn** [comment]
@lee1043 @acordonez 
Could you help review this pull request when you have time? It is not urgent because this branch should be merged right before our paper is submitted.

I have a question. In the JSON output structure (see below), there is a hierarchical structure for seasons and months. Is this structure working for CMEC as well? or Should I update it as a flat structure? 

image

**Comment by acordonez** [comment]
@msahn I'll review this by the end of the week. Can I run these metrics with data that is already in the PMP sample data set?

**Comment by msahn** [comment]
> @msahn Can you add your driver script(s) to setup.py [here](https://github.com/PCMDI/pcmdi_metrics/blob/main/setup.py#L66)?

Thank you for pointing it out. I have added my driver script to setup.py.

**Comment by acordonez** [comment]
> @lee1043 @acordonez Could you help review this pull request when you have time? It is not urgent because this branch should be merged right before our paper is submitted.
> 
> I have a question. In the JSON output structure (see below), there is a hierarchical structure for seasons and months. Is this structure working for CMEC as well? or Should I update it as a flat structure?
> 
> image

@msahn This structure works fine in the CMEC JSONs.

File: pcmdi_metrics/precip_distribution/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Precip distribution metrics
```
Comment (acordonez) [inline pcmdi_metrics/precip_distribution/README.md:1]: Would it be possible to add a little information to the README about what types of input data the driver takes? For example is it daily timeseries, monthly averages, etc, and which variables. Also do you have to run the driver separately for each dataset, or can you run multiple datasets at once?

File: pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py [hunk 2]
```diff
@@ -0,0 +1,1535 @@
+import cdms2 as cdms
+import MV2 as MV
+import cdutil
+import genutil
+import numpy as np
+import glob
+import copy
+import pcmdi_metrics
+import regionmask
```
[the remark this answers is not in the document]
Comment (msahn) [reply pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:9]: How can I check the current dependencies for the PMP? I think the packages for defining regions below need to be additionally added.
- regionmask
- rasterio
- shapely

File: pcmdi_metrics/precip_distribution/README.md (hunk 1 above)
Comment (msahn) [reply pcmdi_metrics/precip_distribution/README.md:1]: Thank you for your suggestions. I have added the information to the README.

File: pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (acordonez) [reply pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:9]: The dependencies for conda-forge are defined here: https://github.com/acordonez/pcmdi_metrics-feedstock/blob/master/recipe/meta.yaml#L30

While shapely is not explicitly defined, it is already being installed in the pcmdi_metrics environment as a dependency for the defined packages. Rasterio and regionmask are not. If you want to see everything that gets installed by default, you can create a pmp environment and use conda list like this:
```
conda create -n new_pmp_env -c conda-forge pcmdi_metrics
conda activate new_pmp_env
conda list
```

When I was testing the IMERG job, I also found that 'netcdf4' needed to be explicitly installed or I'd get an error in xarray. So in total I think we'd need to add 'rasterio', 'regionmask', and 'netcdf4' to meta.yaml.

[the remark this answers is not in the document]
Comment (msahn) [reply pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:9]: Thank you for the information. Adding the three packages sounds good to me.

[the remark this answers is not in the document]
Comment (lee1043) [reply pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py]: @acordonez Are those libraries needs to be added to `conda-env/dev.yml` file as well?  Sorry for my delayed response on this thread.

**Comment by lee1043** [comment]
@msahn I have made some changes to pass code-formatting check from pre-commit.

While I have resolved most of them, followings are still remaining and I need your help to fix them. Could you please take a look and fix them? You will need to do `git pull` first before you make any change in your branch, to avoid any code conflict. 

```
pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:546:5: F841 local variable 'pmax' is assigned to but never used
pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:572:9: F841 local variable 'binrend' is assigned to but never used
pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:1212:5: F841 local variable 'ar6_ocean' is assigned to but never used
pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:1921:5: F841 local variable 'ar6_ocean' is assigned to but never used
pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py:12:43: F821 undefined name 'modpath'
pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py:24:47: F821 undefined name 'res'
pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py:24:80: F821 undefined name 'res'
pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip6.py:12:43: F821 undefined name 'modpath'
pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip6.py:24:47: F821 undefined name 'res'
pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip6.py:24:80: F821 undefined name 'res'
```

**Comment by msahn** [comment]
> @msahn I have made some changes to pass code-formatting check from pre-commit.
> 
> While I have resolved most of them, followings are still remaining and I need your help to fix them. Could you please take a look and fix them? You will need to do `git pull` first before you make any change in your branch, to avoid any code conflict.
> 
> ```
> pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:546:5: F841 local variable 'pmax' is assigned to but never used
> pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:572:9: F841 local variable 'binrend' is assigned to but never used
> pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:1212:5: F841 local variable 'ar6_ocean' is assigned to but never used
> pcmdi_metrics/precip_distribution/lib/lib_precip_distribution.py:1921:5: F841 local variable 'ar6_ocean' is assigned to but never used
> pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py:12:43: F821 undefined name 'modpath'
> pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py:24:47: F821 undefined name 'res'
> pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py:24:80: F821 undefined name 'res'
> pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip6.py:12:43: F821 undefined name 'modpath'
> pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip6.py:24:47: F821 undefined name 'res'
> pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip6.py:24:80: F821 undefined name 'res'
> ```

Thank you very much for your help fixing the code style and format.

I have removed the unused local variables in "lib_precip_distribution.py" but could not fix the issues in "parallel_driver_cmip5.py" and "parallel_driver_cmip6.py". In the codes, ["modpath" and "res" are defined from parameter files](https://github.com/PCMDI/pcmdi_metrics/blob/1d43cab8626f2398bde0380ed0cdacea6a89bf97/pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py#L9-L10), but it seems like the checking algorithm can not recognize it.

**Comment by lee1043** [comment]
> I have removed the unused local variables in "lib_precip_distribution.py" but could not fix the issues in "parallel_driver_cmip5.py" and "parallel_driver_cmip6.py". In the codes, ["modpath" and "res" are defined from parameter files](https://github.com/PCMDI/pcmdi_metrics/blob/1d43cab8626f2398bde0380ed0cdacea6a89bf97/pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py#L9-L10), but it seems like the checking algorithm can not recognize it.

@msahn I merged `parallel_driver_cmip5.py` and `parallel_driver_cmip6.py` into a single file, [`parallel_driver_cmip.py`](https://github.com/PCMDI/pcmdi_metrics/blob/94d8c0ac5f18bd956c04bbee06da730f6b87607b/pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip.py), which reads in the parameter file.

```python -u parallel_driver_cmip.py -p ../param/precip_distribution_params_cmip5.py```

Its usage for cmip5 and cmip6 is briefly documented [here](https://github.com/PCMDI/pcmdi_metrics/blob/94d8c0ac5f18bd956c04bbee06da730f6b87607b/pcmdi_metrics/precip_distribution/scripts_pcmdi/README.md). Could you check if this is okay and work for you?

I also wonder if this setup allows running the code for just one or a few models. If so could you explain how?

**Comment by msahn** [comment]
> > I have removed the unused local variables in "lib_precip_distribution.py" but could not fix the issues in "parallel_driver_cmip5.py" and "parallel_driver_cmip6.py". In the codes, ["modpath" and "res" are defined from parameter files](https://github.com/PCMDI/pcmdi_metrics/blob/1d43cab8626f2398bde0380ed0cdacea6a89bf97/pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip5.py#L9-L10), but it seems like the checking algorithm can not recognize it.
> 
> @msahn I merged `parallel_driver_cmip5.py` and `parallel_driver_cmip6.py` into a single file, [`parallel_driver_cmip.py`](https://github.com/PCMDI/pcmdi_metrics/blob/94d8c0ac5f18bd956c04bbee06da730f6b87607b/pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip.py), which reads in the parameter file.
> 
> `python -u parallel_driver_cmip.py -p ../param/precip_distribution_params_cmip5.py`
> 
> Its usage for cmip5 and cmip6 is briefly documented [here](https://github.com/PCMDI/pcmdi_metrics/blob/94d8c0ac5f18bd956c04bbee06da730f6b87607b/pcmdi_metrics/precip_distribution/scripts_pcmdi/README.md). Could you check if this is okay and work for you?
> 
> I also wonder if this setup allows running the code for just one or a few models. If so could you explain how?

Thank you for updating the code to be more efficient. I have tested it, and there was no issue. I also have added functionality in [parallel_driver_cmip.py](https://github.com/PCMDI/pcmdi_metrics/blob/33aecbf51b7577f833f9c457d212c8afd311ec05/pcmdi_metrics/precip_distribution/scripts_pcmdi/parallel_driver_cmip.py#L17-L21) for running the code for one model using `--mod` option.

**Comment by lee1043** [comment]
After [adding dependencies in dev.yml](https://github.com/PCMDI/pcmdi_metrics/pull/861/commits/5b8e6a0bf966e10e88b660dccce894ceb06cc541) github action's build test get automatically cancelled every time it was triggered. 

@acordonez do you have any idea on this? Should we consider specifying versions of the newly added dependencies?

**Comment by lee1043** [comment]
Note: Below four dependencies were added in `conda-env/dev.yml`. I think I can include them as default for now. 

  - `netcdf4=1.6.0`
  - `regionmask=0.9.0`
  - `rasterio=1.2.10`
  - `shapely=1.8.0`

However, considering they are only used for precip distribution metric, further discussion will be needed to decide whether we want to keeping them as default or optional. 

@acordonez @msahn any comment welcome.

**Comment by acordonez** [comment]
@lee1043 We'll also need to add those as dependencies in the conda feedstock meta.yaml if we want them to be installed with PMP via conda-forge. I've had a lot of trouble with shapely in particular trying to install it after creating a new environment with other packages in it, so installing these at the same time as the other PMP dependencies is probably the best way to go.

**Comment by lee1043** [comment]
> @lee1043 We'll also need to add those as dependencies in the conda feedstock meta.yaml if we want them to be installed with PMP via conda-forge. I've had a lot of trouble with shapely in particular trying to install it after creating a new environment with other packages in it, so installing these at the same time as the other PMP dependencies is probably the best way to go.

@acordonez  good to know! Agreed, thanks for your input!

**Comment by acordonez** [comment]
@lee1043 @msahn Looking at the build failures...I don't see anything wrong, but I'll keep looking.

**Comment by acordonez** [comment]
@lee1043 @msahn I'm stumped about this failure. I've tried rerunning the job a couple of times and it happens in the same step each time. I don't see any warnings or messages related to the new dependencies, but maybe there's some additional step we have to take to include them. Do you think we could ask Jason or Tom?

**Comment by lee1043** [comment]
@acordonez thanks for checking. Same here, I have looked through messages, rerunned build test multiple times with enabling debugging, but it keeps cancels after ~5 min for no obvious region, which is strange. Do you think it is okay we just override it for now...?

**Comment by lee1043** [comment]
By above 2 commits, I confirmed that the build test cancelling was not influenced by newly added dependencies in the dev.yml. With that it may be okay to override...

**Review by lee1043 — APPROVED** [review]
I have confirmed the code works well on my end. Thanks @msahn for the contribution!

**Pull request merged**